mm2src/coins/lp_coins.rs

```diff
@@ -818,6 +822,68 @@ pub struct WatcherReward {
     pub send_contract_reward_on_spend: bool,
 }
 
+/// Enum representing possible variants of swap transaction including secret hash(es)
+#[derive(Debug)]
+pub enum SwapTxTypeWithSecretHash<'a> {
+    /// Legacy protocol transaction
+    TakerOrMakerPayment { maker_secret_hash: &'a [u8] },
+    /// Taker funding transaction
+    TakerFunding { taker_secret_hash: &'a [u8] },
+    /// Maker payment v2 (with immediate refund path)
+    MakerPaymentV2 {
+        maker_secret_hash: &'a [u8],
+        taker_secret_hash: &'a [u8],
+    },
+    /// Taker payment v2
+    TakerPaymentV2 { maker_secret_hash: &'a [u8] },
+}
+
+impl<'a> SwapTxTypeWithSecretHash<'a> {
+    pub fn redeem_script(&self, time_lock: u32, my_public: &Public, other_public: &Public) -> Script {
+        match self {
+            SwapTxTypeWithSecretHash::TakerOrMakerPayment { maker_secret_hash } => {
+                payment_script(time_lock, maker_secret_hash, my_public, other_public)
+            },
+            SwapTxTypeWithSecretHash::TakerFunding { taker_secret_hash } => {
+                swap_proto_v2_scripts::taker_funding_script(time_lock, taker_secret_hash, my_public, other_public)
+            },
+            SwapTxTypeWithSecretHash::MakerPaymentV2 {
+                maker_secret_hash,
+                taker_secret_hash,
+            } => swap_proto_v2_scripts::maker_payment_script(
+                time_lock,
+                maker_secret_hash,
+                taker_secret_hash,
+                my_public,
+                other_public,
+            ),
+            SwapTxTypeWithSecretHash::TakerPaymentV2 { maker_secret_hash } => {
+                swap_proto_v2_scripts::taker_payment_script(time_lock, maker_secret_hash, my_public, other_public)
+            },
+        }
+    }
+
+    pub fn add_op_return_data(&self, op_return_builder: Builder) -> Builder {
+        match self {
+            SwapTxTypeWithSecretHash::TakerOrMakerPayment { maker_secret_hash } => {
+                op_return_builder.push_bytes(maker_secret_hash)
+            },
+            SwapTxTypeWithSecretHash::TakerFunding { taker_secret_hash } => {
+                op_return_builder.push_bytes(taker_secret_hash)
+            },
+            SwapTxTypeWithSecretHash::MakerPaymentV2 {
+                maker_secret_hash,
+                taker_secret_hash,
+            } => op_return_builder
+                .push_bytes(maker_secret_hash)
+                .push_bytes(taker_secret_hash),
+            SwapTxTypeWithSecretHash::TakerPaymentV2 { maker_secret_hash } => {
+                op_return_builder.push_bytes(maker_secret_hash)
+            },
+        }
+    }
```

Comment:
I wasn't able to find a way to implement `as_bytes` for `SwapTxTypeWithSecretHash` because `MakerPaymentV2` variant has 2 fields, which likely can't be merged to another slice without additional allocation. What do you think about`fn op_return_data(&self) -> Vec<u8>`? Or maybe there's a way to still return the data as slice, which I didn't discover? 🙂

Reply:
AFAIK "push_bytes" contains some magical operations so merging maker_secret_hash and taker_secret_hash then using one "push_bytes" wouldn't help. "fn op_return_data(&self) -> Vec<u8>" seems good idea to me :)

Reply:
Done